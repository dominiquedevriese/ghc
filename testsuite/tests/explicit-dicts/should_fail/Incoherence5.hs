{-# LANGUAGE FlexibleInstances #-}
module Incoherence5 where

class SomeClass a where
  someMethod :: a -> a

instance SomeClass a where
  someMethod = id

f :: SomeClass a => a -> a
f = someMethod
    -- Which dictionary will be used here? The one passed by the constraint or
    -- the one defined by the top-level catch-all instance?

-- When passing an explicit dictionary to f, it is not clear that the
-- dictionary will actually be used, because the top-level catch-all instance
-- could be used in f.
test = f ((notDict :: SomeClass a)) True
  where
    notDict = SomeClass.Dict not
