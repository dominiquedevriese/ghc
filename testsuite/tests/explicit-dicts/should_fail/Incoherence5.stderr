
Incoherence5.hs:10:6: warning: [-Wsimplifiable-class-constraints (in -Wdefault)]
    • The constraint ‘SomeClass a’ matches an instance declaration
      instance SomeClass a -- Defined at Incoherence5.hs:7:10
      This makes type inference for inner bindings fragile;
        either use MonoLocalBinds, or simplify it using the instance
    • In the type signature: f :: SomeClass a => a -> a

Incoherence5.hs:18:8: warning: [-Wincoherence (in -Wdefault)]
    • Explicit dictionary application to: ‘SomeClass a’
      is not allowed because of possible incoherence:
        There is a global instance for ‘SomeClass a’
    • In the expression: f notDict True
      In an equation for ‘test’:
          test
            = f notDict True
            where
                notDict = SomeClass.Dict not

<no location info>: error: 
Failing due to -Werror.
